presto-main/src/main/java/com/facebook/presto/sql/gen/RowExpressionPredicateCompiler.java

```diff
@@ -0,0 +1,292 @@
+                throw new PrestoException(COMPILER_ERROR, e);
+            }
+        };
+    }
+
+    private static ParameterizedType generateFilterClassName(Optional<String> classNameSuffix)
+    {
+        return makeClassName(Predicate.class.getSimpleName(), classNameSuffix);
+    }
+
+    private ClassDefinition defineFilterClass(RowExpression filter, int[] inputChannels, CallSiteBinder callSiteBinder, Optional<String> classNameSuffix)
+    {
+        ClassDefinition classDefinition = new ClassDefinition(
+                a(PUBLIC, FINAL),
+                generateFilterClassName(classNameSuffix),
+                type(Object.class),
+                type(Predicate.class));
+
+        CachedInstanceBinder cachedInstanceBinder = new CachedInstanceBinder(classDefinition, callSiteBinder);
+
+        generateFilterMethod(classDefinition, callSiteBinder, cachedInstanceBinder, filter);
+
+        // getInputChannels
+        classDefinition.declareMethod(a(PUBLIC), "getInputChannels", type(int[].class))
+                .getBody()
+                .append(invoke(callSiteBinder.bind(inputChannels, int[].class), "getInputChannels"))
+                .retObject();
+
+        // constructor
+        generateConstructor(classDefinition, cachedInstanceBinder, method -> {});
+
+        return classDefinition;
+    }
+
+    private MethodDefinition generateFilterMethod(
+            ClassDefinition classDefinition,
+            CallSiteBinder callSiteBinder,
+            CachedInstanceBinder cachedInstanceBinder,
+            RowExpression filter)
+    {
+        Map<LambdaDefinitionExpression, LambdaBytecodeGenerator.CompiledLambda> compiledLambdaMap = generateMethodsForLambda(classDefinition, callSiteBinder, cachedInstanceBinder, filter);
+
+        Parameter session = arg("session", ConnectorSession.class);
+        Parameter page = arg("page", Page.class);
+        Parameter position = arg("position", int.class);
+
+        MethodDefinition method = classDefinition.declareMethod(
+                a(PUBLIC),
+                "test",
+                type(boolean.class),
+                ImmutableList.<Parameter>builder()
+                        .add(session)
+                        .add(page)
+                        .add(position)
+                        .build());
+
+        method.comment("Filter: %s", filter.toString());
+
+        Scope scope = method.getScope();
+        BytecodeBlock body = method.getBody();
+
+        declareBlockVariables(filter, page, scope, body);
+
+        Variable wasNullVariable = scope.declareVariable("wasNull", body, constantFalse());
+        RowExpressionCompiler compiler = new RowExpressionCompiler(
+                callSiteBinder,
+                cachedInstanceBinder,
+                fieldReferenceCompiler(callSiteBinder),
+                functionManager,
+                compiledLambdaMap);
+
+        Variable result = scope.declareVariable(boolean.class, "result");
```

Comment:
I might have missed something but `RowExpressionCompiler` is a quite powerful tool to compile any `RowExpression` beyond predicates. So, in theory, `result` could be `Object.class`. Is it possible to make this class a generic compiler rather than only for predicates?

Reply:
@highker It should be possible to change this into a generic expression evaluator, but what's the use case? If a generic version returns an Object, then current use case for the filter would require an undesirable cast to boolean.